Integrate async mode for autograd engine with distributed autograd.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31508 Integrate async mode for autograd engine with distributed autograd.**

This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 with about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

pritamdamania committed:
Integrate async mode for autograd engine with distributed autograd.

This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 without about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 9eb7165:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9eb71659c269863054ab557fbd05e4fb0f4f79f2).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4026537) pytorch_linux_xenial_py2_7_9_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77966834) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77966834))
<details>
<summary>
<code>Dec 20 01:08:55 RuntimeError: test_dataloader failed!</code>
</summary>

```
Dec 20 01:08:55 Ran 62 tests in 88.936s 
Dec 20 01:08:55  
Dec 20 01:08:55 FAILED (failures=1, skipped=8) 
Dec 20 01:08:55  
Dec 20 01:08:55 Generating XML reports... 
Dec 20 01:08:55 Traceback (most recent call last): 
Dec 20 01:08:55   File "test/run_test.py", line 456, in <module> 
Dec 20 01:08:55     main() 
Dec 20 01:08:55   File "test/run_test.py", line 449, in main 
Dec 20 01:08:55     raise RuntimeError(message) 
Dec 20 01:08:55 RuntimeError: test_dataloader failed! 
Dec 20 01:08:55 + cleanup 
Dec 20 01:08:55 + retcode=1 
Dec 20 01:08:55 + set +x 
Dec 20 01:08:55 =================== sccache compilation log =================== 
Dec 20 01:08:55 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Dec 20 01:08:55  
Dec 20 01:08:55 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 20 01:08:55 Compile requests                 46 
Dec 20 01:08:55 Compile requests executed        27 
Dec 20 01:08:55 Cache hits                       20 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

pritamdamania committed:
Update on "Integrate async mode for autograd engine with distributed autograd."


This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 with about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

[ghstack-poisoned]

albanD commented on test/dist_autograd_test.py:
```diff
@@ -1135,7 +1135,9 @@ def test_backward_node_failure(self):
                         wait_until_node_failure(rank)
 
-                with self.assertRaisesRegex(RuntimeError, "Request aborted during client shutdown"):
+                with self.assertRaisesRegex(RuntimeError,
```
Can you add a comment that explains what are the two possible errors here and why both are ok?

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -386,7 +386,11 @@ void GraphTask::set_exception(
+    if (!future_result_->completed()) {
+      future_result_->setError(e.what());
+    } else {
+      TORCH_INTERNAL_ASSERT(future_result_->hasError());
```
Can that happen due to a race condition if two thread set the error at the same time?

albanD replied:
Ho the lock was outside the online view.. Sounds good !

albanD reviewed (commented):
Looks good, just minor things.

pritamdamania committed:
Update on "Integrate async mode for autograd engine with distributed autograd."


This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 with about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

[ghstack-poisoned]

albanD commented:
Can you rebase to make sure the CI pass properly?

pritamdamania committed:
Update on "Integrate async mode for autograd engine with distributed autograd."


This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 with about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Integrate async mode for autograd engine with distributed autograd."


This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 with about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Integrate async mode for autograd engine with distributed autograd."


This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 with about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

[ghstack-poisoned]

pritamdamania87 commented:
@albanD Thanks for the review, looks like CI is clean now.

albanD reviewed (approved):
The autograd part looks good to me. But you may want an opinion from someone more familiar with the async/futures.

jjlilley commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -195,35 +177,40 @@ void DistEngine::computeDependencies(
+  auto futureGrads = engine_.execute_with_graph_task(
+      autogradContext->retrieveGraphTask(), graphRoot);
+
+  futureGrads->addCallback(
```
This function does actual work in the future callback, which is a little sketchy, since the isCompleted() result isn't technically correct anymore (i.e. it can return isCompleted() == true when real work is still happening; downstream consumers like rpc_agent rely on this value being correct for short-cut codepaths)

But, this should be ok (and saves creating a Future), because it's purely an internal/private function, and you control the two use cases inside this file.

But it might be worth a short "warning" comment here, since the Future being returned here won't work in general, e.g. (you can probably make up something better than this :) )

  // Warning: the returned future is for internal use only, since it does
  // processing in the callback, hence isCompleted() is going to be set earlier
  // than the actual logical completion of the future.

Or maybe just create a separate future, though that's not strictly necessary.

jjlilley replied:
Looks great, thanks for separating this out!

(btw, I do agree with the decision in Future<> to set completed=true at the beginning of markCompleted(), since I don't think we want to deal with callbacks fiddling with the future, and all the odd corner cases that could happen there as a result :) )

jjlilley commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -195,35 +177,40 @@ void DistEngine::computeDependencies(
+          // compute this gradient. Also accumulate grads only for
+          // AccumulateGrad function.
+          if (grads[i].defined() &&
+              dynamic_cast<AccumulateGrad*>(outputEdges[i].function.get())) {
```
doesn't need to be solved in this change: When I see the dynamic_cast<>, I wonder a bit about how we handling throwing exceptions in our Future callbacks.

Right now, we're basically ignoring them, which probably mostly works. Not sure what's better here (there's not really a good way to change the existing completed future to an error), but it might be worth thinking about...

jjlilley replied:
Agree that this case is going to be completely benign.

jjlilley reviewed (commented):
Glad to see more futurization of autograd!
Minor comments below:

jjlilley commented:
FTR, re: futures - I'm good with landing this as-is, if you prefer (though a few-line comment-only follow-up change would be welcome).

From a Future<> standpoint, since people tend to copy-and-paste code around, I mostly want to avoid too much proliferation of Future<> code that does actual work necessary to the future completion in the callback (which will cause bugs if downstream code relies on isCompleted() being correct). Your case above is safe, but having a warning might help prevent people from copying and using in cases where it's less safe.

pritamdamania committed:
Update on "Integrate async mode for autograd engine with distributed autograd."


This PR builds on top of https://github.com/pytorch/pytorch/pull/31230
to ensure that distributed autograd doesn't block an RPC thread anymore during
the backward pass.

I've also added a unit test where all ranks hammer rank 0 with about 60
backward calls (which would cause a deadlock earlier), but now such a test
passes without any issues.

Differential Revision: [D19188749](https://our.internmc.facebook.com/intern/diff/D19188749/)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
@jjlilley Addressed your comments, could you take another look? Thanks!
